### DISCUSSION TITLE
Nested fallback routing

### DISCUSSION THREAD

**Discussion by kylewlacy** [tokio-rs/axum#558] [General]
I noticed that, at least after #408, Axum doesn't allow having a `.nest` router that has a fallback (well as of 0.3.4 it's possible to add, but it never gets routed to-- because of #488 / #529 this looks like it will become a panic soon).

We have a use case where having a nested fallback is useful, specifically we have a structure like this:

- `/v1`
  - `/swagger`
    - `/schema.json` → Return fixed response
    - Other requests → Route to `StaticDir`
  - (... other routes ...)

So, if `.nest()` could take a `Router` with a `.fallback()`, then we could easily model the above. Without it, it would start to get messy. For now, we came up with the following simple workaround: using [`service_method_routing::any`](https://docs.rs/axum/0.3.4/axum/routing/service_method_routing/fn.any.html) to wrap the `Router`. Here's what it looks like in code:

```rust
let swagger = axum::Router::new()
    .route("/schema.json", routing::get(|| async { "schema.json" }))
    .fallback(
        tower_http::services::ServeDir::new(path)
            .handle_error(|_| (StatusCode::INTERNAL_SERVER_ERROR, "Internal server error")),
    );

let router = axum::Router::new()
    .nest("/v1/swagger", routing::service_method_routing::any(swagger)) // <---- Wrapping the router
    .fallback(fallback_handler.into_service());
```

(Here's a full working example: https://github.com/kylewlacy/test-axum-nested-routing)

---

So there's the context basically. For purposes of discussion, here are 2 questions to start:

1. Is the above use of `service_method_routing::any` to wrap a router _intended_? That is, can we rely on the above behavior indefinitely?
2. Would it be possible to get Axum to do the more intuitive thing here, and use a nested router's fallback? As a new user of Axum, it was pretty surprising to me that this didn't work out of the box (but I also understand it's likely a tradeoff that was made while balancing competing priorities).

**Comment by jplatte** [comment]
I think the main / only reason nested fallbacks aren't supported is that matchit doesn't support having both a `/foo/bar` and a `/foo/*rest` route. If this could be supported there, it should be super easy to translate nested fallbacks to it.

**Reply by davidpdrsn** [reply]
Yes I believe thats correct.
